include/itkVkMultiResolutionPyramidImageFilter.hxx

```diff
@@ -0,0 +1,210 @@
+/*=========================================================================
+ *
+ *  Copyright NumFOCUS
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#ifndef itkVkMultiResolutionPyramidImageFilter_hxx
+#define itkVkMultiResolutionPyramidImageFilter_hxx
+
+#include "itkCastImageFilter.h"
+#include "itkDiscreteGaussianImageFilter.h"
+#include "itkGaussianOperator.h"
+#include "itkMacro.h"
+#include "itkResampleImageFilter.h"
+#include "itkShrinkImageFilter.h"
+#include "itkIdentityTransform.h"
+
+#include "itkMath.h"
+
+namespace itk
+{
+template <typename TInputImage, typename TOutputImage>
+void
+VkMultiResolutionPyramidImageFilter<TInputImage, TOutputImage>::GenerateData()
+{
+  // Mostly reimplements MultiResolutionPyramidImageFilter::GenerateData
+
+  // Get the input and output pointers
+  InputImageConstPointer inputPtr = this->GetInput();
+
+  // Create caster, smoother and resampleShrinker filters
+  using CasterType = CastImageFilter<TInputImage, TOutputImage>;
+
+  using ImageToImageType = ImageToImageFilter<TOutputImage, TOutputImage>;
+  using ResampleShrinkerType = ResampleImageFilter<TOutputImage, TOutputImage>;
+  using ShrinkerType = ShrinkImageFilter<TOutputImage, TOutputImage>;
+
+  auto                               caster = CasterType::New();
+  auto                               spatialSmoother = SpatialSmootherType::New();
+  auto                               fftSmoother = FFTSmootherType::New();
```

Comment:
Perhaps loop through all `GetUseFFT(this->GetKernelRadius(ilevel))` values and check whether `spatialSmoother` will ever by used, and likewise for `fftSmoother`.  If we can avoid `New()` and subsequent initialization for an unused smoother that could be decently big win for time and memory.

Reply:
I've made both smoothers members of the pyramid filter itself. Please see discussion on the most recent comment below.